packages/webview_flutter/lib/webview_flutter.dart

```diff
@@ -307,85 +307,51 @@ class _WebViewState extends State<WebView> {
   }
 }
 
-Set<String> _extractChannelNames(Set<JavascriptChannel> channels) {
-  final Set<String> channelNames = channels == null
-      // TODO(iskakaushik): Remove this when collection literals makes it to stable.
-      // ignore: prefer_collection_literals
-      ? Set<String>()
-      : channels.map((JavascriptChannel channel) => channel.name).toSet();
-  return channelNames;
+CreationParams _creationParamsfromWidget(WebView widget) {
+  return CreationParams(
+    initialUrl: widget.initialUrl,
+    webSettings: _webSettingsFromWidget(widget),
+    javascriptChannelNames:
+    _extractChannelNames(widget.javascriptChannels),
+  );
 }
 
-class _CreationParams {
-  _CreationParams(
-      {this.initialUrl, this.settings, this.javascriptChannelNames});
-
-  static _CreationParams fromWidget(WebView widget) {
-    return _CreationParams(
-      initialUrl: widget.initialUrl,
-      settings: _WebSettings.fromWidget(widget),
-      javascriptChannelNames:
-          _extractChannelNames(widget.javascriptChannels).toList(),
-    );
-  }
-
-  final String initialUrl;
-
-  final _WebSettings settings;
-
-  final List<String> javascriptChannelNames;
-
-  Map<String, dynamic> toMap() {
-    return <String, dynamic>{
-      'initialUrl': initialUrl,
-      'settings': settings.toMap(),
-      'javascriptChannelNames': javascriptChannelNames,
-    };
-  }
+WebSettings _webSettingsFromWidget(WebView widget) {
+  return WebSettings(
+    javascriptMode: widget.javascriptMode,
+    hasNavigationDelegate: widget.navigationDelegate != null,
+    debuggingEnabled: widget.debuggingEnabled,
+  );
 }
 
-class _WebSettings {
-  _WebSettings({
-    this.javascriptMode,
-    this.hasNavigationDelegate,
-    this.debuggingEnabled,
-  });
-
-  static _WebSettings fromWidget(WebView widget) {
-    return _WebSettings(
-      javascriptMode: widget.javascriptMode,
-      hasNavigationDelegate: widget.navigationDelegate != null,
-      debuggingEnabled: widget.debuggingEnabled,
-    );
+WebSettings _webSettingsUpdate(WebSettings currentValue, WebSettings newValue) {
+  JavascriptMode javascriptMode;
+  bool hasNavigationDelegate;
+  bool debuggingEnabled;
+  if (currentValue.javascriptMode != newValue.javascriptMode) {
```

Comment:
if the newValue.javascriptMode is null and the currentValue.javascriptMode is non-null, we are setting the old value to null.  Same as below. It seems to violate the comment here:
```  
/// Updates the webview settings.
///
/// Any non null field in `settings` will be set as the new setting value.
/// All null fields in `settings` are ignored.
```

Reply:
+1

Reply:
The fields in `currentValue` should never be null (the initial value built by _WebSettingFromWidget can never have null values).

I made it clear in the method comment and added assertions.

Reply:
I was thinking if the fields in newValue is null and this will set the corresponding fields in the currentValue to null. But the dart doc in the `updateSetting` in the WebViewSettings class mentioned that the null fields should be ignored.

Reply:
After offline discussion with @amirh; we confirmed that the fields in both currentValue and newValue could never be null because they are both created by _webSettingsFromWidget which takes the setting from the `WebView` widget, which should never be null. Can possibly add assertions in the code but it will create a lot of code overhead.

Reply:
Added more assertions here

Reply:
On the platform side, at least for the platforms we've been dealing with so far, all these "settings" are updated by imperative calls, e.g on Android we call [setJavaScriptEnabled](https://developer.android.com/reference/android/webkit/WebSettings.html#setJavaScriptEnabled(boolean)) to update the "javascriptMode".
We do not want to call [setJavaScriptEnabled](https://developer.android.com/reference/android/webkit/WebSettings.html#setJavaScriptEnabled(boolean)) every time the widget is rebuilt, so we're keeping track of the previous `javascriptMode` value we had the last time the widget was built, and only when we see that the value has changed we call `setJavaScriptEnabled`.

The `WebSettings` class groups these kind of settings. We track the previous `WebSettings` value, and  when the widget is rebuilt this method(`_webSettingsUpdate`) computes a `WebSettings` instance that only includes the values that have changed and has to be updated on the platform side. This value is then passed to `WebViewPlatform.updateSettings` which ignores the null fields and updates the platform webview with the new values for all non-null fields in WebSettings.

Reply:
I see, so you're nullfying all the unchanged fields on purpose because `WebViewPlatform.updateSettings` expects it that way. I guess the method name confused me, it's more like `nullfyUnchangedValues` which is a horrible name I agree. Anyway, thanks for the detailed explanation.

Reply:
Renamed to `_clearUnchangedWebSettings`